### PR TITLE
Add support for custom esprima options; fixes #903

### LINKED ISSUE #903: Ability to control the options passed to esprima.
Esprima (at least for ES6) has a `tolerant` parser option that allows it to parse not-strictly-spec-compliant syntax. This is useful in some cases. For example, I only use ES6 in a transpiled-to-ES5 context, and ES6 only allows `import`s at the top level of a module, but the transpiler (6to5) doesn't have that restriction (because CommonJS's `require` has no such restriction). Allowing such imports makes conditional imports and imports with fallbacks if the module is unavailable, possible.
Thus, I'm requesting that JSCS add an option that controls Esprima's `tolerant` flag, so that JSCS can style check such ES6 code instead of immediately dying with a parsing error like it currently does.
(See also https://code.google.com/p/esprima/issues/detail?id=619)

<bountysource-plugin>

---
Want to back this issue? **[Place a bounty on it!](https://www.bountysource.com/issues/7678251-es6-add-an-option-to-enable-esprima-s-tolerant-flag?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by cvrebert**
Work-in-progress attempt at fixing #903.
This seems to pass through the options properly. The 2 new tests currently fail because `esprima-harmony-jscs` is outdated and doesn't include https://github.com/ariya/esprima/commit/b1dd29e985b353648a6e0fa8b9f852cd2db228ac (Should I file an issue at https://github.com/jscs-dev/esprima-harmony for that?)

### CODE REVIEW

**Comment by mikesherov** [comment]
@cvrebert I'll update Esprima-harmony. Thanks for contributing!

Please add documentation for this feature, which I assume is config only and not available via CLI.

I leave the rest of my feedback as line comments.

**Comment by cvrebert** [comment]
@mikesherov Cool, I'll add docs. Please LMK when the new esprima-harmony is ready.

**Comment by cvrebert** [comment]
Added docs to the Overview. Let me know if more need to be added somewhere else.

**Comment by mikesherov** [comment]
New version is ready: `1.1.0-tolerate-import`

**Comment by mikesherov** [comment]
@cvrebert you'll need to rebase, as I extracted parsing into it's own function.

File: lib/string-checker.js [hunk 1]
```diff
@@ -84,18 +85,19 @@ StringChecker.prototype = {
     checkString: function(str, filename) {
         filename = filename || 'input';
         str = str.replace(/^#!?[^\n]+\n/gm, '');
+        var esprimaOptions = extend({
```
Comment (mikesherov) [inline lib/string-checker.js]: `... extend({}, this._configuration.getEsprimaOptions(), {...` so that user can't overwrite the required options and that the passed in options aren't modified.

Comment (mikesherov) [inline lib/string-checker.js]: Also, `_extend` isn't documented Node API. Might be better just to use a `forin` here.

File: test/string-checker.js [hunk 2]
```diff
@@ -200,6 +200,30 @@ describe('modules/string-checker', function() {
         });
     });
 
+    describe('esprima options', function() {
+        it('uses custom esprima options when provided to the constructor', function() {
+            checker = new Checker({ esnext: true, esprimaOptions: { tolerant: true } });
+            checker.registerDefaultRules();
+
+            var errors = checker.checkString('(function () {import * as foo from "bar"; return foo; })()');
+            console.log(errors);
```
Comment (mikesherov) [inline test/string-checker.js]: remove

File: test/string-checker.js [hunk 3]
```diff
@@ -200,6 +200,30 @@ describe('modules/string-checker', function() {
...
+            assert(errors.isEmpty());
```
Comment (mikesherov) [inline test/string-checker.js]: combine with previous line.

File: test/string-checker.js [hunk 4]
```diff
@@ -200,6 +200,30 @@ describe('modules/string-checker', function() {
...
+        it('uses custom esprima options when set in the config', function() {
+            checker = new Checker();
+            checker.registerDefaultRules();
+            checker.configure({ esnext: true, esprimaOptions: { tolerant: true } });
+
+            var errors = checker.checkString('(function () { import * as bar from "bar"; return bar; })();');
+            // Make sure that multiple checks don't fail
+            var errors2 = checker.checkString('(function () { import * as bar from "foo"; return bar; })()');
+            console.log(errors);
+            assert(errors.isEmpty());
```
Comment (mikesherov) [inline test/string-checker.js]: combine with `:218`

File: test/string-checker.js [hunk 5]
```diff
@@ -200,6 +200,30 @@ describe('modules/string-checker', function() {
...
+            assert(errors2.isEmpty());
```
Comment (mikesherov) [inline test/string-checker.js]: combine with `:220`

File: OVERVIEW.md [hunk 6]
```diff
@@ -214,6 +214,20 @@ Value: `true`
...
+Custom `options` to be passed to `esprima.parse(code, options)`
+
+Type: `Object`
+
+Default: `{}`
+
+#### Example
+
+```js
+"esprimaOptions": { "tolerant": true }
```
Comment (mikesherov) [inline OVERVIEW.md]: we should consider making this part of the default.

**Comment by mikesherov** [comment]
@cvrebert thanks again. Let me know when you've made the necessary changes. Also, as far as tests go, it would be good to not really on the fact that esprima:tolerant now tolerates non-top level imports.
A better test would use a custom esprima who's parse function is a spy, and ensure the correct combined options are passed in.

Thanks again for contributing!

**Comment by cvrebert** [comment]
Note to self: The method extraction was done in 0253717989b475999e2deea19f124ff1cbb502fd

File: lib/config/configuration.js [hunk 7]
```diff
@@ -38,6 +38,7 @@ function Configuration() {
     this._esnextEnabled = false;
     this._es3Enabled = true;
     this._esprima = null;
+    this._esprimaOptions = null;
```
Comment (mikesherov) [inline lib/config/configuration.js]: `this._esprimaOptions = {};`

File: OVERVIEW.md (hunk 6 above)
Comment (cvrebert) [inline OVERVIEW.md]: Okay, made it the default.

**Comment by cvrebert** [comment]
@mikesherov Addressed your feedback.

File: test/string-checker.js [hunk 8]
```diff
@@ -207,6 +207,32 @@ describe('modules/string-checker', function() {
...
+                error.description = options.foobar === 'qux' ? customDescription : 'should never happen';
+                error.lineNumber = 1;
+                error.column = 0;
+
+                throw error;
+            }
+        };
+
+        it('uses custom esprima options when set in the config', function() {
+            checker = new Checker({ esprima: customEsprima });
```
Comment (mikesherov) [inline test/string-checker.js]: before this, Add `var sinon = require('sinon');` to the top of this module, then: `sinon.spy(customEsprima, 'parse');`. Also, don't pass `esprima` into the construction, pass it in via `checker.configure`, which is the preferred way.

File: test/string-checker.js [hunk 9]
```diff
@@ -207,6 +207,32 @@ describe('modules/string-checker', function() {
...
+            checker.registerDefaultRules();
+            checker.configure({ esprimaOptions: { foobar: 'qux' } });
+
+            var errors = checker.checkString('import { foo } from "bar";');
```
Comment (mikesherov) [inline test/string-checker.js]: Just `checker.checkString('import { foo } from "bar";');`

File: test/string-checker.js [hunk 10]
```diff
@@ -207,6 +207,32 @@ describe('modules/string-checker', function() {
...
+            var error = errors.getErrorList()[0];
```
Comment (mikesherov) [inline test/string-checker.js]: remove retrieving and checking the error.

Comment (cvrebert) [inline test/string-checker.js]: But then the test doesn't properly test anything..

File: test/string-checker.js [hunk 11]
```diff
@@ -207,6 +207,32 @@ describe('modules/string-checker', function() {
...
+
+            assert(error.rule === 'parseError');
+            assert(error.message === customDescription);
```
Comment (mikesherov) [inline test/string-checker.js]: Instead `assert(costumEsprima.parse.calledWith(expectedOptions))` which checks literally that the right options were called.

Once you have that set up, write a few more tests: 

1. one where you provide no options and see tolerant is used.
1. one where you provide options that aren't overwriteable.
1. one where you provide tolerant and see it's overriden.
1. one where you provide arbitrary options see that they're used.

P.S. see http://sinonjs.org/docs/#spies if you need more info on spies.

File: OVERVIEW.md (hunk 6 above)
Comment (mikesherov) [inline OVERVIEW.md]: Thanks!

**Comment by mikesherov** [comment]
@cvrebert please rebase once more, and address remaining feedback. Thanks again for sticking with this!

File: test/string-checker.js (hunk 10 above)
Comment (mikesherov) [inline test/string-checker.js]: Read my notes below. It tests that the correct options are passed into the parse function by spying on the parse function.

File: package.json [hunk 12]
```diff
@@ -63,7 +63,7 @@
     "colors": "~1.0.3",
     "commander": "~2.5.0",
     "esprima": "~1.2.2",
-    "esprima-harmony-jscs": "1.1.0-regex-token-fix",
+    "esprima-harmony-jscs": "1.1.0-tolerate-import",
```
Comment (mikesherov) [inline package.json]: Perhaps this update should move into a separate PR to address #952 ?

File: test/string-checker.js (hunk 8 above)
Comment (cvrebert) [inline test/string-checker.js]: Can't use `configure`; it only accepts strings, not functions.

Comment (mikesherov) [inline test/string-checker.js]: Ah, good point! Comment withdrawn.

**Comment by cvrebert** [comment]
@mikesherov Addressed 2nd round of feedback.

**Comment by mikesherov** [comment]
@cvrebert thanks again! You're almost there. Coverage went down because you need to explicity write a test in `test/config/configuration` for `getEsprimaOptions`. 

Once that's in, this looks good to land. Thanks again for the work you put in here.

**Comment by cvrebert** [comment]
@mikesherov Added config tests.

**Comment by mikesherov** [comment]
@cvrebert thanks again! I just landed this with a few tweaks:

1. pulled out the tolerate-import update to be addressed in a different PR.
2. removed some redundancy in the tests.

Otherwise, landed as is. Thanks again a bunch for the effort here!
